dataprep/clean/clean_duplication_utils.py

```diff
@@ -0,0 +1,323 @@
+        do_merge
+            Boolean values indicating whether a cluster should be merged.
+            If do_merge[i] == True then merge the i-th cluster.
+        """
+        df_name, col = self._df_name, "'" + self._col + "'"
+        replace_calls = []
+        for idx, cluster in enumerate(cluster_page):
+            cluster_repr = "'" + new_values[idx] + "'"
+            if do_merge[idx]:
+                cluster_vals = ["'" + cluster_val + "'" for cluster_val, _ in cluster]
+                code = (
+                    f"{df_name}[{col}] = {df_name}[{col}].replace"
+                    f"([{', '.join(cluster_vals)}], {cluster_repr})"
+                )
+                replace_calls.append(code)
+
+        return "\n" + "\n".join(replace_calls)
+
+    def live_export_code(
+        self, cluster_page: pd.Series, do_merge: List[bool], new_values: List[str]
+    ) -> None:
+        """
+        Create DataFrame replace calls to merge clusters and output
+        them into the jupyter notebook.
+        """
+        code = self._create_replace_calls(cluster_page, do_merge, new_values)
+        encoded_code = (base64.b64encode(str.encode(code))).decode()
+        code = """
+            var ind = IPython.notebook.get_selected_index();
+            var cell = IPython.notebook.get_cell(ind);
+            var text = cell.get_text();
+            cell.set_text(text.concat(atob("{0}")));
+        """.format(
+            encoded_code
+        )
+        display(Javascript(code))
+
+    def execute_merge_code(
+        self,
+        cluster_page: pd.Series,
+        do_merge: List[bool],
+        new_values: List[str],
+    ) -> None:
+        """
+        Merge the clusters in the DataFrame.
+        """
+        for idx, cluster in enumerate(cluster_page):
+            cluster_repr = new_values[idx]
+            if do_merge[idx]:
+                self._df.loc[:, self._col] = self._df[self._col].replace(
+                    [cluster_val for cluster_val, _ in cluster], cluster_repr
+                )
+
+    def final_df(self) -> None:
+        """
+        Displays a DataFrame with the final values in the next notebook cell.
+        """
+        code = "# dataframe with cleaned string values\ndf_clean"
+        encoded_code = (base64.b64encode(str.encode(code))).decode()
+        code = """
+            IPython.notebook.kernel.execute("df_clean = {1}.copy()");
+            var code = IPython.notebook.insert_cell_below('code');
+            code.set_text(atob("{0}"));
+            code.execute();
+        """.format(
+            encoded_code, self._df_name
+        )
+        display(Javascript(code))
+
+    def set_cluster_params(self, ngram: int, radius: int, block_size: int) -> None:
+        """
+        Set clustering parameters.
+        """
+        self._ngram = ngram
+        self._radius = radius
+        self._block_size = block_size
+
+
+def _levenshtein_distance(str1: str, str2: str) -> int:
```

Comment:
Have you looked into computing the levenshtein distance in C? https://github.com/ztane/python-Levenshtein

Reply:
I'll make this change, I wanted to avoid adding a dependency but it's probably worth it for the performance boost.

Reply:
Okay. I think it would also be possible to write it in Cython and store the compiled C file in our repo. E.g., something like the `query_integral` function in this [wordcloud](https://github.com/amueller/word_cloud/blob/master/wordcloud/query_integral_image.pyx) library. If you find a suitable library to take on as a dependency, I think we can do that. Or if you want to look into whether we can write it in Cython, I think that's okay too. We can also do this optimization later, but like you said I think it would be worth the performance boost.

Reply:
I added the library you linked for now. I'll try to look into removing the dependency by writing it in Cython in another PR.